include/visitors/code_gen.h

```diff
@@ -2,30 +2,545 @@
+            {
+                while_stmt->accept(this);
+            }
+
+            if (auto for_stmt = dynamic_cast<ForStmt *>(stmt.get()))
+            {
+                for_stmt->accept(this);
+            }
+
+            if (auto return_stmt = dynamic_cast<ReturnStmt *>(stmt.get()))
+            {
+                return_stmt->accept(this);
+            }
+
+            if (auto break_stmt = dynamic_cast<BreakStmt *>(stmt.get()))
+            {
+                break_stmt->accept(this);
+            }
+
+            if (auto continue_stmt = dynamic_cast<ContinueStmt *>(stmt.get()))
+            {
+                continue_stmt->accept(this);
+            }
+        }
+
+        create_entry_point();
+
+        BinaryenModulePrint(this->mod);
+
+        BinaryenModuleAllocateAndWriteResult result =
+            BinaryenModuleAllocateAndWrite(this->mod, nullptr);
+
+        if (!result.binary || result.binaryBytes == 0)
+        {
+            std::cerr << "failed to serialize" << std::endl;
+            return;
+        }
+
+        std::string filename = "output.wasm";
+        std::ofstream file(filename, std::ios::binary);
+        if (file.is_open())
+        {
+            file.write(static_cast<char *>(result.binary), result.binaryBytes);
+            file.close();
+            std::cout << "wasm module written to " << filename << std::endl;
+        }
+        else
+        {
+            std::cerr << "failed to open file" << filename << std::endl;
+        }
+
+        free(result.binary);
+
+        this->environment->pop_env();
+    }
+
+    int allocate_local()
+    {
+        int index = this->local_index++;
+        return index;
+    }
+
+    BinaryenType from_bird_type(Token token)
+    {
+        if (token.lexeme == "bool")
+            return BinaryenTypeInt32(); // no i1?
+        else if (token.lexeme == "int")
+            return BinaryenTypeInt32();
+        else if (token.lexeme == "float")
+            return BinaryenTypeFloat64();
+        else if (token.lexeme == "void")
+            return BinaryenTypeNone();
+        else if (token.lexeme == "str")
+            return BinaryenTypeInt32(); // ptr
+        else
+            throw BirdException("invalid type");
+    }
+
+    void visit_block(Block *block)
+    {
+        // throw BirdException("Implement visit_block");
+    }
+
+    void visit_decl_stmt(DeclStmt *decl_stmt)
+    {
+        decl_stmt->value->accept(this);
+        BinaryenExpressionRef initializer_value = this->stack.pop();
+
+        BinaryenType type = decl_stmt->type_identifier.has_value()
+                                ? from_bird_type(decl_stmt->type_identifier.value())
+                                : BinaryenExpressionGetType(initializer_value);
+
+        int index = allocate_local(); // locals defined by index instead of name
+        locals.push_back(type);
+
+        BinaryenExpressionRef set_local = BinaryenLocalSet(this->mod, index, initializer_value);
```

Comment:
I think a local set is the right thing here, but I do want to point out that the data will only persist during a function's lifetime. Eventually, when we want something to persist over the span of multiple functions, we will use the load and store functions.